Feature/ollama request customization

ptitjes opened the pull request:
This PR add the ability to customize Ollama chat requests as discussed in #295.

@aozherelyeva I had to make two modification to the `OllamaTestFixture` and I hope it's OK:
- Added a check to pull the llama3.2 model in the container, because by default the image that you suggested in TESTING.md doesn't come with the llama3.2 model. Now it works without touching anything (provided you have docker obviously).
- Exposed the `baseUrl` to allow some tests to initialize a custom `OllamaClient` if required.

---

#### Type of the change
- [x] New feature
- [ ] Bug fix
- [ ] Documentation fix

#### Checklist for all pull requests
- [x] The pull request has a description of the proposed change
- [x] I read the [Contributing Guidelines](https://github.com/JetBrains/koog/blob/main/CONTRIBUTING.md) before opening the pull request
- [x] The pull request uses **`develop`** as the base branch
- [x] Tests for the changes have been added
- [x] All new and existing tests passed

##### Additional steps for pull requests adding a new feature
- [x] An issue describing the proposed change exists
- [x] The pull request includes a link to the issue
- [x] The change was discussed and approved in the issue
- [ ] Docs have been added / updated

ptitjes commented:
And I had forgotten to add Kdocs. That's now fixed.

Ololoshechkin commented on prompt/prompt-executor/prompt-executor-clients/prompt-executor-ollama-client/src/commonMain/kotlin/ai/koog/prompt/executor/ollama/client/OllamaClient.kt:
```diff
@@ -42,7 +76,8 @@ public class OllamaClient(
     baseClient: HttpClient = HttpClient(engineFactoryProvider()),
     timeoutConfig: ConnectionTimeoutConfig = ConnectionTimeoutConfig(),
-    private val clock: Clock = Clock.System
+    private val requestBuilderAction: (OllamaRequestBuilder.(prompt: Prompt, model: LLModel) -> Unit)? = null,
+    private val clock: Clock = Clock.System,
```
Also if you make this lambda the last parameter -- you would be able to write 
```
val client = OllamaClient(...) {
   seed = 0
   numCtx = 100
}
```

Also, could you please clarify once again -- why do you need to configure it with a lambda and not just pass a data object with this parameter, ex:

```
val client = OllamaClient(..., requestOptions = OllamaRequestOptions(seed = 0, numCtx = 100))
```

devcrocod replied:
I agree with @Ololoshechkin that using a lambda as the last parameter is better.
I understand your motivation to make it more explicit, but in my opinion, code readability is more important, and this change makes it worse.

Also, if possible, it’s better to use only one receive parameter inside the lambda. it’s not yet clear to me why both `prompt` and `model` are needed there

ptitjes replied:
> It would be sufficient for me to just set the num_ctx to the max possible.

I think that would impair a lot of users, because not everybody has great hardware...

For example, I run my agents on my tiny AMD 5700 with 8GB of VRAM. I usually use `hf.co/mradermacher/ToolACE-2-Llama-3.1-8B-i1-GGUF:Q4_K_M` for tool related stuff. It is a llama3.1 fine-tune, and thus supports a 128K (131072) context length. If I set `OLLAMA_CONTEXT_LENGTH=131072`, this mean 16384.00 MiB for the KV cache in addition to the 4685.30 MiB of the model. I get all of the model and cache loaded into RAM and using CPU. When I try to keep the context relatively small, I get at some parts of the model and cache loaded in the VRAM and using the GPU, and the generation is very much faster... So, automatically setting `num_ctx` to the max context length of the model would make Koog unusable for me.

Also, newer models like llama4 scout support up to 10M context length. Automatically setting `num_ctx` to the max would make any request (even with the smallest context) unmanageable even on the greatest hardware. (In q8_0 KV cache type, Ollama stores 16K context per GiB, which means 640 GiB of (V)RAM needed!!! :eyes: )

ptitjes replied:
> The context calculation has prompt and model, but does this prompt also include the tools formulation? I'm thinking if we have a list of tools with names and descriptions, this fills the context as well. Either from Koog if it builds the description, or inside ollama if it expands the model template.

No, the prompt don't contain the tool formulation. Indeed, Ollama is expanding that itself in the model template. So, in any case, we can only do a token count estimation.

But isn't that better than doing nothing?

Ololoshechkin replied:
Other than that, might be a great idea to make the main kdoc above the `OllamaClient` much longer and explain all these details, and even provide a few code samples how to use this new configuration lambda (feel free to check the kdoc of `AgentMemory.Feature` or `EventHandler.Feature` for instance -- we already have quite long explanations in Kdocs :)). Also it's nice for LLMs to learn on these things once they read them :)

ptitjes replied:
> adding an example to the examples module that would showcase the Ollama with fine-tuned dynamic context lenght based on the prompt and model -- will be ideal. Plus, if you could put some textual explanations there in this example, that would benefit everyone's understanding and enrich the knowledgebase around Koog. WDYT? :)

Yeah, definitely.

ptitjes commented:
Rebased on develop and applied requested changes:
- made `requestBuilderAction` the last (non-nullable) constructor parameter
- checked that the response is the same when seed=0

ptitjes commented:
I'll have to rebase because the llama3.2 pull of https://github.com/JetBrains/koog/pull/335/commits/1012015883cd4bf9adb3f76c4c0efacf84eaf385 was merged in #371.

ptitjes commented:
Rebased on develop and removed the commit which @aozherelyeva also implemented in develop.

devcrocod commented on prompt/prompt-executor/prompt-executor-clients/prompt-executor-ollama-client/src/commonMain/kotlin/ai/koog/prompt/executor/ollama/client/dto/OllamaModels.kt:
```diff
@@ -71,6 +71,9 @@ internal data class OllamaChatRequestDTO(
     internal data class Options(
         val temperature: Double? = null,
+        val seed: Int? = null,
+        @SerialName("num_ctx") val numCtx: Int? = null,
```
Doesn’t the Ollama json config assume the use of `snake_case` flag, so that we don’t have to use `@SerialName`?

devcrocod replied:
Ideally, all of this should be caught by tests, so I don’t see a big issue with it
Besides, we already use `namingStrategy` in other clients
So far, the only real advantage is a slight reduction in code

I don’t think it’s critical, since we can remove it or add it to the other clients at any time if needed

devcrocod reviewed (commented):
I’m worried that Ollama’s implementation is drifting further and further from the others.

Also, I think it would be better to add additional functions or constructors to creating so that I can store the request parameters in a variable and pass them that way. It’s more convenient when writing custom wrapper functions on top of koog

ptitjes commented on integration-tests/src/jvmMain/kotlin/ai/koog/integration/tests/OllamaTestFixture.kt:
```diff
@@ -18,6 +18,7 @@ class OllamaTestFixture {
 
     private lateinit var ollamaContainer: GenericContainer<*>
 
+    lateinit var baseUrl: String
```
Because it is initialised in `setUp` (so after construction) from the ollama test container's host and port, just like `client` and `executor`.

Would you prefer that I change it's type to `String?`?

devcrocod replied:
Oh, I see
I thought the URL was already set at the time the class was initialized

devcrocod commented on prompt/prompt-executor/prompt-executor-clients/prompt-executor-ollama-client/src/commonMain/kotlin/ai/koog/prompt/executor/ollama/client/OllamaClient.kt:
```diff
@@ -42,7 +78,8 @@ public class OllamaClient(
     timeoutConfig: ConnectionTimeoutConfig = ConnectionTimeoutConfig(),
-    private val clock: Clock = Clock.System
+    private val clock: Clock = Clock.System,
+    private val requestBuilderAction: OllamaRequestBuilder.(prompt: Prompt, model: LLModel) -> Unit = { _, _ -> },
```
> I don't understand what you mean here. This requestBuilderAction lambda is exactly called inside execute passing the very prompt and model provided to the execute() call. How come is there "no guarantee that the model here will be the same as the one used in execute"?

Yea, i missed that
you're absolutely right

> I can set model as an immutable val field of OllamaRequestBuilder, but that would be even way worse here, because then OllamaRequestBuilder would both have mutable fields to build the request and an immutable field to get information about the model... This would make it confusing and hardly discoverable for the user, which is bad design IMO.

true

Ololoshechkin replied:
@ptitjes I believe we can just move all not-so-much-popular params to `OllamaRequestBuilder` receiver. Then, `model` will be always available (those who know they know :) ), while `prompt` will be the only param of the lambda. We usually follow such approach in the DSLs in general.

WDYT?

ptitjes replied:
@Ololoshechkin I started implementing something like this, but I am really wondering if the `maxResponseTokenCount` shouldn't simply be in `LLMParams` as all the proprietary-models behind APIs also have some form of max generated tokens.

If we decide to make this one an `LLMParams`, then I would only compute the max context window length in the strategy.

Does this make sense? WDYT?

ptitjes commented:
Let's put this on hold for now and take some time discussing a better solution. I can live with configuring bigger context sizes directly on Ollama for some time. I just hope we can agree on an acceptable design in the next Koog releases, because this is at the expense of memory and performance.

Ololoshechkin commented:
> I have faith that in the future small LLMs with great performances will become more and more common. That is good for people and good for the climate. IMO Koog has to support these use-cases.

Totally agree with you here, @ptitjes !

ptitjes commented:
> I don't think we can make it for 0.3.0 already, unfortunately :(
>
> But in fact -- adding ollama as a first thing and as experimental in 0.3.* is fine -- we don't have to be waiting for 0.4.0.

Yes, I prefer we get an almost right design, rather than rush it.

> Experimental annotation -- no, we currently don't have it. So might be a great time to add it.

I definitely think that we should have such an annotation and start being more conservative (maybe not from 0.3 or 0.4, but soon) on the new APIs we introduce. Otherwise that might bite us later.

ptitjes commented:
@Ololoshechkin I really like your `ContextWindowStrategy` interface proposal, coming with some built-in implementations and a sensible default.

I will rework this PR in this way. I need a few days to think about the appropriate built-in implementations to cover the main use-cases.

Some initial thoughts though:

1. I wood remove `seed` there, as it is completely unrelated. But at the same time, I think it's only useful for tests so I wouldn't mind having that as a simple constructor parameter of `OllamaClient`. If you are OK with that, I will do a separate PR with only this and the corresponding integration test.

2. This `ContextWindowStrategy` would be something specific to Ollama, as I believe none of the other clients need anything similar. As far as I can see, all the proprietary models behind API handle this by themselves. So that wouldn't make much sense for those, right?

3. I am convinced we should add new attributes to `LLModel`, namely `contextLength : Long` and `embeddingLength : Long`. That would be really helpful to inform the context summarization strategies (and also this `ContextWindowStrategy` as a result). WDYT?

ptitjes commented:
So, finally coming back to this. I rewrote everything following @Ololoshechkin's advice.

I defined a `ContextWindowStrategy` with a single method that is responsible to compute the context length value to send to Ollama as `num_ctx`. This value computed value can be null in case the user wishes to let the Ollama server deal with that. (This is fully documented in the KDocs of `ContextWindowStrategy`.)

I implemented three basic strategies:
- `None`, that simply returns `null` to let the Ollama server deal with this.
- `Fixed`, that returns a constant value. If this value is greater than what the current model supports, then we fallback to the maximum context length supported by the model (and log a warning).
- `FitPrompt`, that computes a context length value so that the current prompt fits in the context window. This strategy uses either the given prompt tokenizer or the last reported token usage in the prompt. The strategy also ensured that computed context length is a multiple of the given `granularity` (in order to avoid too many context length changes, and thus model reloads by Ollama) and is coerced between a given minimum context length and the max supported context length of the model.

I had to move `PromptTokenizer` and its implementation to the `prompt-tokenizer` module. (I don't know why it was lost near the Tokenizer agent feature.)

I added unit tests that check that the correct `num_ctx` value is sent to Ollama, by using a MockHttpClient. I considered that using integration tests would be really difficult (given we would have to produce very long prompts) and so not worth the cost.

aozherelyeva reviewed (approved):
The part regarding the `OllamaTestFixture` is more than okay, thanks for the contribution!
